src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/14.diff.sql

```diff
@@ -0,0 +1,537 @@
+     (
+         ResourceChangeTypeId tinyint NOT NULL,
+         Name nvarchar(50) NOT NULL,
+         CONSTRAINT PK_ResourceChangeType PRIMARY KEY CLUSTERED (ResourceChangeTypeId),
+         CONSTRAINT UQ_ResourceChangeType_Name UNIQUE NONCLUSTERED (Name)
+     ) 
+     ON [PRIMARY]
+END
+GO
+
+IF (EXISTS (SELECT 1 FROM sys.tables WHERE name = 'ResourceChangeType') AND
+     NOT EXISTS (SELECT 1 FROM dbo.ResourceChangeType WHERE ResourceChangeTypeId = 0))
+BEGIN
+     INSERT dbo.ResourceChangeType (ResourceChangeTypeId, Name) VALUES (0, N'Creation')
+END
+GO
+
+IF (EXISTS (SELECT 1 FROM sys.tables WHERE name = 'ResourceChangeType') AND
+     NOT EXISTS (SELECT 1 FROM dbo.ResourceChangeType WHERE ResourceChangeTypeId = 1))
+BEGIN
+     INSERT dbo.ResourceChangeType (ResourceChangeTypeId, Name) VALUES (1, N'Update')
+END
+GO
+
+IF (EXISTS (SELECT 1 FROM sys.tables WHERE name = 'ResourceChangeType') AND
+     NOT EXISTS (SELECT 1 FROM dbo.ResourceChangeType WHERE ResourceChangeTypeId = 2))
+BEGIN
+     INSERT dbo.ResourceChangeType (ResourceChangeTypeId, Name) VALUES (2, N'Deletion')
+END
+GO
+
+/*************************************************************
+    Stored procedures for capturing and fetching resource changes
+**************************************************************/
+--
+-- STORED PROCEDURE
+--     CaptureResourceChanges
+--
+-- DESCRIPTION
+--     Inserts resource change data
+--
+-- PARAMETERS
+--     @isDeleted
+--         * Whether this resource marks the resource as deleted.
+--     @version
+--         * The version of the resource being written
+--     @resourceId
+--         * The resource ID
+--     @resourceTypeId
+--         * The ID of the resource type
+--
+-- RETURN VALUE
+--     It does not return a value.
+--
+CREATE OR ALTER PROCEDURE dbo.CaptureResourceChanges
+    @isDeleted bit, 
+    @version int, 
+    @resourceId varchar(64),
+    @resourceTypeId smallint
+AS
+BEGIN
+    DECLARE @changeType SMALLINT
+    IF (@isDeleted = 1) BEGIN
+        SET @changeType = 2    /* DELETION */
+    END
+    ELSE BEGIN
+        IF (@version = 1) BEGIN
+            SET @changeType = 0 /* CREATION */
+        END
+        ELSE BEGIN
+            SET @changeType = 1 /* UPDATE */
+        END
+    END
+
+    INSERT INTO dbo.ResourceChangeData
+        (ResourceId, ResourceTypeId, ResourceVersion, ResourceChangeTypeId)
+    VALUES
+        (@resourceId, @resourceTypeId, @version, @changeType)
+END
+GO
+
+--
+-- STORED PROCEDURE
+--     FetchResourceChanges
+--
+-- DESCRIPTION
+--     Returns the number of resource change records from startId. The start id is inclusive.
+--
+-- PARAMETERS
+--     @startId
+--         * The start id of resource change records to fetch.
+--     @pageSize
+--         * The page size for fetching resource change records.
+--
+-- RETURN VALUE
+--     Resource change data rows.
+--
+CREATE OR ALTER PROCEDURE dbo.FetchResourceChanges
+    @startId bigint, 
+    @pageSize int
```

Comment:
I would add `@DataCount int OUTPUT` parameter and set it to @@ROWCOUNT value, so you always know how much data you getting back.

Reply:
Good point. But, there is no business logic dependent on row counts, and don't see many benefits at this moment. For simplicity, not adding the output parameter. We can use the size of the list in C# if needed.

Reply:
It let you know how big of collection you expect back, so you don't have to reallocate memory for list resize during reading of collection. It's not business driven, but it's for performance, 

I'm talking instead of:
```
var list = new List<Class>();
while (reader.Read()){
list.Add (new Class())
}
```
you can do

```
var list = new List<Class>((int)command.Parameter["DataCount"]);
while (reader.Read()){
list.Add (new Class())
}
```